### PR TITLE
Introduces an Equalize preprocessing layer

### PR DESCRIPTION
**Pull request by LukeWood**
I'd like to discuss the "usefulness" of this style of layer.  

This layer is one of the sub-pieces of the RandAugment technique.  By itself, this layer is probably not very useful.  That being said, we need it for both RandAugment and AutoAugment.

This layer doesn't feel very useful by itself, but if we're going to use it in two of our other augmentation layers then we may as well offer it as an end user facing API.  My only concern would be API pollution.

Thoughts?  @fchollet @qlzh727

### CODE REVIEW

**Comment by bhack** [comment]
Few question:
- Do we think we will need to use and Equalize (or any other subprocessing) as a layer? Are we interested to let the users to cherry pick the single individual preprocessing operation to compose their model?
- Do we need to use Equalize only as private API for other more complex layers or e.g. for visualization APIs?
- Do we need to use the Equalize op as a public API like [`tfa.image.equalize`](https://www.tensorflow.org/addons/api_docs/python/tfa/image/equalize)?

File: keras_cv/layers/preprocessing/equalize.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+
+class Equalize(tf.keras.layers.Layer):
+    """Equalize performs histogram equalization on a channel-wise basis."""
+
+    def __init__(self, rate=1.0, **kwargs):
+        super().__init__(**kwargs)
+
+    @staticmethod
+    def scale_channel(image, c):
```
Comment (fchollet) [inline keras_cv/layers/preprocessing/equalize.py:24]: Use fully-spelled out variable names, e.g. `channel_index`

File: keras_cv/layers/preprocessing/equalize.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        image = tf.cast(image[..., c], tf.int32)
+        # Compute the histogram of the image channel.
+        histo = tf.histogram_fixed_width(image, [0, 255], nbins=256)
```
Comment (fchollet) [inline keras_cv/layers/preprocessing/equalize.py:27]: histogram

File: keras_cv/layers/preprocessing/equalize.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+        # For the purposes of computing the step, filter out the nonzeros.
+        nonzero = tf.where(tf.not_equal(histo, 0))
+        nonzero_histo = tf.reshape(tf.gather(histo, nonzero), [-1])
+        step = (tf.reduce_sum(nonzero_histo) - nonzero_histo[-1]) // 255
+
+        def build_mapping(histo, step):
+            # Compute the cumulative sum, shifting by step // 2
+            # and then normalization by step.
+            lut = (tf.cumsum(histo) + (step // 2)) // step
```
Comment (fchollet) [inline keras_cv/layers/preprocessing/equalize.py:37]: lut?

File: keras_cv/layers/preprocessing/equalize.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import tensorflow as tf
+
+
+class Equalize(tf.keras.layers.Layer):
+    """Equalize performs histogram equalization on a channel-wise basis."""
```
Comment (fchollet) [inline keras_cv/layers/preprocessing/equalize.py:18]: Add `Args` section

File: keras_cv/layers/preprocessing/equalize.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        result = tf.cond(
+            tf.equal(step, 0),
+            lambda: image,
+            lambda: tf.gather(build_mapping(histo, step), image),
+        )
+
+        return tf.cast(result, tf.uint8)
+
+    def call(self, images):
+        """call method for Equalize
```
Comment (fchollet) [inline keras_cv/layers/preprocessing/equalize.py:55]: Because docs are generated from the class docstring, `call()` should be documented in a `Call args` section in the class docstring. See e.g. https://keras.io/api/layers/recurrent_layers/lstm/

File: keras_cv/layers/preprocessing/equalize.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+        Args:
+            images: Tensor of type int, pixels in range [0, 255], in RGB format.
+        """
+        # Assumes RGB for now.  Scales each channel independently
+        # and then stacks the result.
+        r = tf.map_fn(lambda x: Equalize.scale_channel(x, 0), images)
+        g = tf.map_fn(lambda x: Equalize.scale_channel(x, 1), images)
+        b = tf.map_fn(lambda x: Equalize.scale_channel(x, 2), images)
+        images = tf.stack([r, g, b], -1)
```
Comment (fchollet) [inline keras_cv/layers/preprocessing/equalize.py:65]: To maximize readability, avoid positional values with an ambiguous meaning, e.g. here use `axis=-1`

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

File: keras_cv/layers/preprocessing/equalize.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import tensorflow as tf
+
+
+class Equalize(tf.keras.layers.Layer):
+    """Equalize performs histogram equalization on a channel-wise basis."""
+
+    def __init__(self, rate=1.0, **kwargs):
```
Comment (qlzh727) [inline keras_cv/layers/preprocessing/equalize.py]: why the rate is not saved as a property?

File: keras_cv/layers/preprocessing/equalize.py (hunk 1 above)
Comment (qlzh727) [reply keras_cv/layers/preprocessing/equalize.py]: agreed. Also if we expect this to be static method, probably just not put in the class, and move it to the module level.

File: keras_cv/layers/preprocessing/equalize.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+        super().__init__(**kwargs)
+
+    @staticmethod
+    def scale_channel(image, c):
+        image = tf.cast(image[..., c], tf.int32)
```
Comment (qlzh727) [inline keras_cv/layers/preprocessing/equalize.py]: This is assuming channel last format. Is it safe to have this assumption?

Comment (qlzh727) [reply keras_cv/layers/preprocessing/equalize.py]: Also the cast seems interesting here. This is expecting the image to be in a certain range? eg [0, 255]? Might want to put this in the docstring.

File: keras_cv/layers/preprocessing/equalize.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+        # If step is zero, return the original image.  Otherwise, build
+        # lut from the full histogram and step and then index from it.
+        result = tf.cond(
+            tf.equal(step, 0),
+            lambda: image,
+            lambda: tf.gather(build_mapping(histo, step), image),
+        )
+
+        return tf.cast(result, tf.uint8)
```
Comment (qlzh727) [inline keras_cv/layers/preprocessing/equalize.py:57]: Should we cast the output image back to its original dtype?

File: keras_cv/layers/preprocessing/equalize.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+    def call(self, images):
+        """call method for Equalize
+
+        Args:
+            images: Tensor of type int, pixels in range [0, 255], in RGB format.
+        """
+        # Assumes RGB for now.  Scales each channel independently
+        # and then stacks the result.
+        r = tf.map_fn(lambda x: Equalize.scale_channel(x, 0), images)
```
Comment (qlzh727) [inline keras_cv/layers/preprocessing/equalize.py:62]: This seems to be able to vectorized in future for performance improvement. Might worth a TODO here.

**Review by qlzh727 — CHANGES_REQUESTED** [review]
Should have a unit test for this, as well as docstring for how to use this layer.

**Comment by LukeWood** [comment]
> Few open questions:
> 
> * Do we think we will need to use and Equalize (or any other subprocessing) as a layer? Are we interested to let the users to cherry pick the single individual preprocessing operation to compose their model?
> * Do we need to use Equalize only as a private API for other more complex layers or e.g. for visualization APIs?
> * Do we need to use the Equalize op as a public API like [`tfa.image.equalize`](https://www.tensorflow.org/addons/api_docs/python/tfa/image/equalize)?

So, this is part of an overall bigger effort to make a RandAugment layer composed of sub-preprocessing layers.  Instead of writing it with a mixture of ops and layers (because some of the augmentations would need to be layers anyways),  I'm offering them all as layers.

I admit in the case of equalize this is not super useful in isolation, but it will allow for cleaner composability/consistency across the API.

**Comment by qlzh727** [comment]
> Few open questions:
> 
> * Do we think we will need to use and Equalize (or any other subprocessing) as a layer? Are we interested to let the users to cherry pick the single individual preprocessing operation to compose their model?
I think individual layer will be useful, even just for user who want to do a one-off process and see the outcome as experiment.
Combining multiple layer into one augment layer will certainly be useful as well, which I think is proposed in the RandAugment layer.
> * Do we need to use Equalize only as a private API for other more complex layers or e.g. for visualization APIs?
The non-layer op API is useful IMO, since it is self-contained and has a very clear API (equalize an input image). Layer is probably just a container/wrapper of that op, so that it can be used by other layer/model.

> * Do we need to use the Equalize op as a public API like [`tfa.image.equalize`](https://www.tensorflow.org/addons/api_docs/python/tfa/image/equalize)?
Ah, should we port/use the tfa.image.equalize to avoid duplication?

**Comment by LukeWood** [comment]
>> Do we need to use Equalize only as a private API for other more complex layers or e.g. for visualization APIs?
> The non-layer op API is useful IMO, since it is self-contained and has a very clear API (equalize an input image). Layer is probably just a container/wrapper of that op, so that it can be used by other layer/model.
>> Do we need to use the Equalize op as a public API like [tfa.image.equalize](https://www.tensorflow.org/addons/api_docs/python/tfa/image/equalize)?
> Ah, should we port/use the tfa.image.equalize to avoid duplication?

Let's see if there's a request for it.  We can also update the KPL to work on single images if needed.  For now, if users need it they can relate on tfa.image.equalize.

**Comment by bhack** [comment]
> Ah, should we port/use the tfa.image.equalize to avoid duplication?

@qlzh727  It was just a general question using this  PR as a good example as we merged the `CALL_FOR_CONTRIBUTIONS.md` without incorporating clarifications and other open topics we had in PR review. 
If we are going ahead with the general approach to expose these kind of ops in the public API  other then the layer I think that we could remove these from TF Addons and more in general to not collect anymore PRs with similar ops in TFA. 
We have already shear, sharpen etc.. in TFA but I already see there in the public issue queue here...
I think it is important to understand if we want to go with public or private at the ops level as, also in the private only case, we are going to "duplicate" the implementations.

File: keras_cv/layers/preprocessing/equalize.py (hunk 10 above)
Comment (bhack) [reply keras_cv/layers/preprocessing/equalize.py]: Also in cases where we don't have issues with the limits of `tf.vectorized_map`[https://www.tensorflow.org/api_docs/python/tf/vectorized_map] we could try to evaluate that as a temp solution (with a TODO for full vectorization)

File: keras_cv/layers/preprocessing/equalize.py [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+
+    Args:
+        bins: number of bins to use in histogram equalization.
+
+    Usage:
+        ```
+        equalize = Equalize()
+
+        (images, labels), _ = tf.keras.data.cifar10.load_data()
+        # Note that images are an int32 Tensor with values in the range [0, 255]
```
Comment (fchollet) [inline keras_cv/layers/preprocessing/equalize.py:28]: That's actually surprising, I'd have expected uint8?

File: keras_cv/layers/preprocessing/equalize.py [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+
+        return tf.cast(result, dtype)
+
+    def call(self, images):
+        if not images.dtype.is_integer:
+            raise ValueError(
+                "Equalize.call(images) expects images to be a "
+                "Tensor of type int, with values in range [0, 255]. "
+                f"got images.dtype={images.dtype}."
+            )
```
Comment (fchollet) [inline keras_cv/layers/preprocessing/equalize.py:86]: @mattdangerw please see code above -- I'd like to get your thoughts on restricting an image processing layer to integer inputs -- I recall we had some discussion about whether KPL should or should not output floats (which would be more consistent with other layers, which cast their inputs to `compute_dtype`). Do you remember what the conclusion was?

If we end up doing that, then to retain the ability to chain layers, we should not require a particular input dtype here.

Comment (mattdangerw) [reply keras_cv/layers/preprocessing/equalize.py:86]: I think it's OK to do, it will just expose a bug in Sequential. Something like

```
model = tf.keras.Sequential([
  Equalize(),
  ...
])
model.build((128, 128, 3))
```

Will trigger the error above, when sequential tries to trace with a float input. We need to do something like adding an input spec to preprocessing layers like this and respecting the input spec in sequential to fix.

File: keras_cv/layers/preprocessing/equalize.py [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+        (images, labels), _ = tf.keras.datasets.cifar10.load_data()
+        # Note that images are an int8 Tensor with values in the range [0, 255]
+        images = equalize(images)
+        ```
+
+    Call arguments:
+        images: Tensor of type int, pixels in range [0, 255], in RGB format.
+    """
+
+    def __init__(self, bins=256, **kwargs):
```
Comment (mattdangerw) [inline keras_cv/layers/preprocessing/equalize.py:36]: What happens when this value is >256?

File: keras_cv/layers/preprocessing/equalize.py [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+
+
+class Equalize(tf.keras.layers.Layer):
+    """Equalize performs histogram equalization on a channel-wise basis.
+
+    Args:
+        bins: number of bins to use in histogram equalization.
+
+    Usage:
+        ```
```
Comment (fchollet) [inline keras_cv/layers/preprocessing/equalize.py:24]: Add python tag, unindent section, add line break aboive

File: keras_cv/layers/preprocessing/equalize.py [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import tensorflow as tf
+
+
+class Equalize(tf.keras.layers.Layer):
+    """Equalize performs histogram equalization on a channel-wise basis.
+
+    Args:
+        bins: number of bins to use in histogram equalization.
```
Comment (fchollet) [inline keras_cv/layers/preprocessing/equalize.py:21]: Add type description (integer)

File: keras_cv/layers/preprocessing/equalize.py [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+        ```
+        equalize = Equalize()
+
+        (images, labels), _ = tf.keras.datasets.cifar10.load_data()
+        # Note that images are an int8 Tensor with values in the range [0, 255]
+        images = equalize(images)
+        ```
+
+    Call arguments:
+        images: Tensor of type int, pixels in range [0, 255], in RGB format.
```
Comment (fchollet) [inline keras_cv/layers/preprocessing/equalize.py:33]: "of pixels"?

Please specify allowable shapes

File: keras_cv/layers/preprocessing/equalize.py (hunk 12 above)
Comment (fchollet) [reply keras_cv/layers/preprocessing/equalize.py:86]: This is problematic since preprocessing layers are likely to be used in `Sequential`... but we'll fix it independently of this specific layer.

File: keras_cv/layers/preprocessing/equalize_test.py [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import tensorflow as tf
+
+from keras_cv.layers.preprocessing.equalize import Equalize
+
+
+class EqualizeTest(tf.test.TestCase):
```
Comment (fchollet) [inline keras_cv/layers/preprocessing/equalize_test.py:19]: We should consider using `layer_test` from Keras to cover standard layer features (including serialization)

Comment (fchollet) [reply keras_cv/layers/preprocessing/equalize_test.py:19]: On this note, we could export `layer_test` as an `keras.__internal__` API so as to guarantee API stability

Comment (qlzh727) [reply keras_cv/layers/preprocessing/equalize_test.py:19]: Yea, I will send a cl for that. addon was using layer_test as well.

Comment (qlzh727) [reply keras_cv/layers/preprocessing/equalize_test.py:19]: Sending https://github.com/keras-team/keras/pull/16021.

**Comment by fchollet** [comment]
A note about naming, now that I think about it: to be consistent with the names of other preprocessing layers, we should use nouns, not verbs, e.g. `Equalization`.

Elsewhere, we use `Rotation`, not `Rotate`, `Normalization`, not `Normalize`, `Vectorization`, not `Vectorize`, `Convolution`, not `Convolve`, etc.

File: keras_cv/layers/preprocessing/equalize.py [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import tensorflow as tf
+
+
+class Equalize(tf.keras.layers.Layer):
```
Comment (fchollet) [inline keras_cv/layers/preprocessing/equalize.py:17]: Please rename to either `Equalization` or `Equalizer` for naming consistency (no strong opinion between the two...)

Comment (LukeWood) [reply keras_cv/layers/preprocessing/equalize.py]: I like Equalization.  Let's go with that.

**Review by fchollet — APPROVED** [review]
